### PR TITLE
PP-1041: moms cannot communicate with one another in a cloud configuration when cloud nodes resolve each other's hostnames to IP addresses not known to the PBS server/comm

### PR DESCRIPTION
**Pull request by bremanandjk**
#### Issue-ID

* **[PP-1041](https://pbspro.atlassian.net/browse/PP-1041)**

#### Problem description
* Earlier PBS_LEAF_NAME would accept only one value. The problem with that was that if the name of the interfaces was not mapped to all the addresses in the host (say via /etc/hosts or dns) then those interface/ips were not registered to PBS_COMM/PBS_SERVER.

#### Cause / Analysis
* As PBS_LEAF_NAME accepts one value, it would resolve that name and find the ip address associated with that name. 

#### Solution description
* Now, we can specify multiple names (comma separated) to the variable PBS_LEAF_NAME. Also if this variable is not set, then we detect all the ip addresses of the machine, even if they do not reverse resolve to a name. All these IP’s are registered with the comm

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/106561543/PP-1040+moms+cannot+communicate+with+one+another+in+a+cloud+configuration+when+cloud+nodes+resolve+each+other+s+hostnames+to+IP+addresses+not+known+to+the+PBS+server+comm)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by bremanandjk** [comment]
Travis centOS build is a success; For openSUSE it's an infrastructure issue.
There is work in progress to fix CI issues.

File: configure [hunk 1]
```diff
@@ -16398,6 +16398,14 @@ fi
...
 $as_echo "#define PYTHON /**/" >>confdefs.h
 
+  if test "x$with_python" == "x"; then :
+
+cat >>confdefs.h <<_ACEOF
+#define SYSTEM_PYTHON_PATH "$PYTHON"
+_ACEOF
+
+
+fi
```
Comment (subhasisb) [inline configure]: Why are we making this change? Is it related to this defect?

File: src/cmds/scripts/pbs_init.d.in [hunk 2]
```diff
@@ -105,10 +105,10 @@ update_pids() {
 }
 
 # lc_host_name - convert host name into lower case short host name
-
+# also handle multiple names in PBS_LEAF_NAME
 lc_host_name() 
 {
-  echo $1 | cut -d: -f1 | cut -d. -f1 | sed -e "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
+  echo $1 | cut -d, -f1 | cut -d: -f1 | cut -d. -f1 | sed -e "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_init.d.in]: Please add a comment to state that PBS_LEAF_NAME is now of the format: **host:port,host:port** so the following code is to cut on , (to get the first host) and then : to parse our the port

File: src/include/pbs_config.h.in [hunk 3]
```diff
@@ -632,6 +632,9 @@
 /* Defined when swig is available */
 #undef SWIG
 
+/* Python executable path */
+#undef SYSTEM_PYTHON_PATH
+
```
Comment (subhasisb) [inline src/include/pbs_config.h.in]: Why this change in this bug fix?

File: src/lib/Liblog/pbs_log.c [hunk 4]
```diff
@@ -332,6 +333,63 @@ log_init(void)
...
+	ni = (struct net_if_info*)malloc(sizeof(struct net_if_info));
+
+	memset(msg, '\0', sizeof(msg));
+	memset(temp, '\0', sizeof(temp));
+
+	get_if_info(ni, msg);
+
+	curr = ni;
+
+	if(strlen(msg)){ /* Adding error message to log */
```
Comment (subhasisb) [inline src/lib/Liblog/pbs_log.c]: Format: Space between ) and { 

in several lines after this as well.

File: configure (hunk 1 above)
Comment (subhasisb) [reply configure]: BTW, is configure part of the checked-in source? I think not.

File: src/include/pbs_config.h.in (hunk 3 above)
Comment (subhasisb) [reply src/include/pbs_config.h.in]: Is this is generated file or part of the src tree?

File: src/lib/Libnet/Makefile.in [hunk 5]
```diff
@@ -195,7 +195,8 @@ libnet_a_LIBADD =
 am_libnet_a_OBJECTS = libnet_a-get_hostaddr.$(OBJEXT) \
```
Comment (subhasisb) [inline src/lib/Libnet/Makefile.in]: Why are we checking in Makefile.in

**Comment by subhasisb** [comment]
@bremanandjk Please run pbs_comm, pbs_server and pbs_mom under valgrind and with all the following cases:

1. Default installation of PBS with 2 hosts with single IP in each host.
2. Default installation of PBS with 2 hosts with multiple ip addresses in each host.
3. On multi-IP hosts, setup PBS to use only one IP address via PBS_LEAF_NAME
4. On hosts using 3 IP addresses, setup PBS to use only 2 of the 3 via PBS_LEAF_NAME
5. Setup on host to report 32 IP addresses and run PBS without PBS_LEAF_NAME.

Please analyze the valgrind outputs for any leaks/corruptions.

File: configure (hunk 1 above)
Comment (bremanandjk) [reply configure]: In our 14.1 release branch, we used to keep GNU auto-generated files.

File: src/include/pbs_config.h.in (hunk 3 above)
Comment (bremanandjk) [reply src/include/pbs_config.h.in]: yes, It is one of the GNU autotool generated files.  Its not related to this bug fix. I ran autogen and included the changes to retain consistency with branch.

File: configure (hunk 1 above)
Comment (mike0042) [reply configure]: We should not retain auto-generated files in any branch.

**Comment by bremanandjk** [comment]
Thanks @subhasisb for your review comments and test suggestions, 
I have tested all the specified scenarios and attached the test results at Jira ticket.

File: configure (hunk 1 above)
Comment (bremanandjk) [reply configure]: Thanks for your suggestion @mike0042. 
Please guide me how can I adhere this, 
Shall I *ignore* auto-generated files in this PR ?  (or) Shall i raise another PR for *removal* of all other auto generated files?

Comment (subhasisb) [reply configure]: @bremanandjk it is fine to leave it the way this branch was before. Thanks for explaining.

File: src/include/pbs_config.h.in (hunk 3 above)
Comment (subhasisb) [reply src/include/pbs_config.h.in]: Sounds good

File: src/lib/Libnet/Makefile.in (hunk 5 above)
Comment (subhasisb) [reply src/lib/Libnet/Makefile.in]: Okay - i see that this branch did not have the required git ignores and so these generated files were being checked in. I know that this is fixed in 18.1 (or master), so I am not worried either way for 14.1.x. Thanks for explaining.

**Comment by subhasisb** [comment]
Thanks for the thorough valgrinding @bremanandjk

**Comment by subhasisb** [comment]
@bremanandjk please squash your commits (and rebase if necessary). You also need another reviewer to sign off on the code changes.

**Comment by bremanandjk** [comment]
Thanks for your approval @subhasisb 
I have squashed my commits to one and added a sign to it.

File: configure (hunk 1 above)
Comment (mike0042) [reply configure]: @bremanandjk there are two things we should be doing...
1. Keeping the list of files in .gitignore up to date, and
2. Remove any auto-generated files we find checked in to the repo.

**Comment by subhasisb** [comment]
@bremanandjk all looks good. We need another secondary reviewer as well. Thanks.

**Comment by suresh-thelkar** [comment]
@subhasisb ,@bremanandjk
I will do the secondary review for this PR.

**Comment by subhasisb** [comment]
@bremanandjk per @mike0042 suggestion, let's remove the auto-generated files from the repo.

File: configure (hunk 1 above)
Comment (bremanandjk) [reply configure]: Thanks, @mike0042  I have implemented your comments.

**Comment by bremanandjk** [comment]
@subhasisb, As per review comment, I have removed auto-generated files.

File: src/lib/Libtpp/tpp_client.c [hunk 6]
```diff
@@ -673,7 +673,8 @@ tpp_init(struct tpp_config *cnf)
 		return -1;
 	}
 
-	TPP_DBPRT(("localhost = %s", tpp_conf->node_name));
+	snprintf(log_buffer, TPP_LOGBUF_SZ, "TPP leaf node names = %s", tpp_conf->node_name);
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c]: Not required since it is a requirement. But no harm adding either. In the case you want to add a check, do it before the snprintf, so if tpp_conf->node_name, you can print an error message and return -1 from this tpp_init() function.

File: src/lib/Libtpp/tpp_dis.c [hunk 7]
```diff
@@ -741,8 +741,9 @@ DIS_tpp_destroy(int fd)
 int
 set_tpp_config(struct pbs_config *pbs_conf,
 	struct tpp_config *tpp_conf,
-	char *nodename,
-	int port, char *r,
+	char *nodenames,
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_dis.c]: This is actually an input, cannot initialize to NULL. In fact there is error checking to abort the routine if it is NULL

File: src/lib/Libtpp/tpp_dis.c [hunk 8]
```diff
@@ -819,18 +824,45 @@ set_tpp_config(struct pbs_config *pbs_conf,
...
+		nm = mk_hostname(token, port);
+		if (!nm) {
+			snprintf(log_buffer, TPP_LOGBUF_SZ, "Failed to make node name");
+			fprintf(stderr, "%s\n", log_buffer);
+			tpp_log_func(LOG_CRIT, NULL, log_buffer);
+			return -1;
+		}
+
+		hlen = strlen(nm);
+		if ((tmp = realloc(formatted_names, len + hlen + 2)) == NULL) { /* 2 for command and null char */
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_dis.c]: yes, but it returns NULL when it fails and if you store it in the same variable ie. formatted_names, you just lost the pointer to you allocated memory block...and you cannot even free it anymore.

File: src/lib/Libtpp/tpp_dis.c [hunk 9]
```diff
@@ -819,18 +824,45 @@ set_tpp_config(struct pbs_config *pbs_conf,
...
+			tpp_log_func(LOG_CRIT, NULL, log_buffer);
+			return -1;
+		}
+
+		formatted_names = tmp;
+
+		if (len == 0) {
+			strcpy(formatted_names, nm);
+		} else {
+			strcat(formatted_names, ",");
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_dis.c]: Not required since exact space required is allocated by doing strlen before itself.

File: src/lib/Libtpp/tpp_util.c [hunk 10]
```diff
@@ -1047,13 +1047,12 @@ char *
...
-	if (!node_name)
-		return NULL;
-	if (port == -1)
-		sprintf(node_name, "%s", host);
-	else
-		sprintf(node_name, "%s:%d", host, port);
-
+	if (node_name) {
+		if (strchr(host, ':') || port == -1)
+			strcpy(node_name, host);
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_util.c]: Nope. Not required, since exactly required space is allocated before

File: src/lib/Libtpp/tpp_util.c [hunk 11]
```diff
@@ -1047,13 +1047,12 @@ char *
...
+		else
+			sprintf(node_name, "%s:%d", host, port);
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_util.c]: Nope. Not required, since exactly required space is allocated before

File: src/lib/Liblog/pbs_log.c [hunk 12]
```diff
@@ -332,6 +333,63 @@ log_init(void)
...
+
+	memset(msg, '\0', sizeof(msg));
+	memset(temp, '\0', sizeof(temp));
+
+	get_if_info(ni, msg);
+
+	curr = ni;
+
+	if(strlen(msg)) { /* Adding error message to log */
+		strncpy(tbuf, msg, LOG_BUF_SIZE);
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Liblog/pbs_log.c]: The other way is to use LOG_BUF_SIZE - 1 in the strncpy or declare tbuf with size LOG_BUF_SIZE + 1, so that the size asked to copy is always less than the dest buffer size, resulting in a NULL termination always!

File: src/lib/Libtpp/tpp_dis.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/lib/Libtpp/tpp_dis.c]: Thanks for catching this.

**Comment by suresh-thelkar** [comment]
@bremanandjk 
Thanks for making the code changes. I sign off.

**Comment by bremanandjk** [comment]
Thanks, @subhasisb & @suresh-thelkar for sign-off. 
Please take it further

File: src/lib/Libtpp/tpp_client.c [hunk 13]
```diff
@@ -673,7 +673,10 @@ tpp_init(struct tpp_config *cnf)
 		return -1;
 	}
 
-	snprintf(log_buffer, TPP_LOGBUF_SZ, "TPP leaf node names = %s", tpp_conf->node_name);
+	if (tpp_conf->node_name)
```
Comment (subhasisb) [inline src/lib/Libtpp/tpp_client.c]: Instead of only printing when node_name is not null, please do the check before and return from the function. So something like this:

f (tpp_conf->node_name == NULL) {
   snprintf(log_buffer, TPP_LOGBUF_SZ, "TPP leaf node name is NULL");
   tpp_log_func(LOG_CRIT, NULL, log_buffer);
   return -1;
}

snprintf(log_buffer, TPP_LOGBUF_SZ, "TPP leaf node names = %s", tpp_conf->node_name);
tpp_log_func(LOG_CRIT, NULL, log_buffer);

Comment (bremanandjk) [reply src/lib/Libtpp/tpp_client.c]: Thanks for your comment @subhasisb 
I have addressed as suggested. please have a look.

**Comment by subhasisb** [comment]
All looks good Brem. I sign off. Please squash your commits in preparation to merge.

**Comment by bremanandjk** [comment]
Thanks @subhasisb 
I have squashed the commits to one. Please take further.
